RecoLocalCalo/EcalRecProducers/plugins/alpaka/EcalUncalibRecHitPhase2WeightsAlgoPortable.dev.cc

```diff
@@ -0,0 +1,80 @@
+#include <alpaka/alpaka.hpp>
+
+#include "DataFormats/EcalDigi/interface/alpaka/EcalDigiPhase2DeviceCollection.h"
+#include "DataFormats/EcalDigi/interface/EcalDataFrame_Ph2.h"
+#include "DataFormats/EcalDigi/interface/EcalDigiCollections.h"
+#include "DataFormats/EcalRecHit/interface/alpaka/EcalUncalibratedRecHitDeviceCollection.h"
+#include "DataFormats/EcalRecHit/interface/EcalUncalibratedRecHit.h"
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/traits.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+
+#include "EcalUncalibRecHitPhase2WeightsAlgoPortable.h"
+#include "EcalUncalibRecHitPhase2WeightsStruct.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE::ecal::weights {
+  using namespace cms::alpakatools;
+
+  class Phase2WeightsKernel {
+  public:
+    ALPAKA_FN_ACC void operator()(Acc1D const& acc,
+                                  EcalUncalibRecHitPhase2Weights const* weightsObj,
+                                  EcalDigiPhase2DeviceCollection::ConstView digisDev,
+                                  EcalUncalibratedRecHitDeviceCollection::View uncalibratedRecHitsDev) const {
+      constexpr int nsamples = ecalPh2::sampleSize;
+      auto const nchannels = digisDev.size();
+      // one thread sets the output collection size scalar
+      if (once_per_grid(acc)) {
+        uncalibratedRecHitsDev.size() = digisDev.size();
+      }
+
+      auto const* weightsdata = weightsObj->weights.data();
+      auto const* timeWeightsdata = weightsObj->timeWeights.data();
+      //divide the grid into uniform elements
+      for (auto tx : uniform_elements(acc, nchannels)) {
+        bool g1 = false;
+        const auto digi = digisDev[tx].data();
+        auto recHit = uncalibratedRecHitsDev[tx];
+        recHit.amplitude() = 0;
+        recHit.jitter() = 0;
+        for (int s = 0; s < nsamples; ++s) {
+          const auto sample = digi[s];
+          const auto trace =
+              (static_cast<float>(ecalLiteDTU::adc(sample))) * ecalPh2::gains[ecalLiteDTU::gainId(sample)];
+          recHit.amplitude() += (trace * weightsdata[s]);
+          recHit.jitter() += (trace * timeWeightsdata[s]);
+          if (ecalLiteDTU::gainId(sample) == 1)
+            g1 = true;
+          recHit.outOfTimeAmplitudes()[s] = 0.;
+        }
+        recHit.amplitudeError() = 1.0f;
+        recHit.id() = digisDev.id()[tx];
+        recHit.flags() = 0;
+        recHit.pedestal() = 0.;
+        recHit.jitterError() = 0.;
+        recHit.chi2() = 0.;
+        recHit.aux() = 0;
+        if (g1) {
+          recHit.flags() = 0x1 << EcalUncalibratedRecHit::kHasSwitchToGain1;
```

Comment:
Is this supposed to be set also if _all_ samples are in gain 1 ?

Reply:
Yes. This flag indicates that the HW has switched from the default gain 10 to gain 1 for at least one of the samples. I do not think having all samples in gain 1 will ever happen anyway.